snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,308 @@
+
+	var processingWeight uint64
+	for nw, lastAccepted := range nodeWeights2Blocks {
+		if sd.processing(lastAccepted) {
+			newProcessingWeight, err := safemath.Add(processingWeight, nw.Weight)
+			if err != nil {
+				sd.log.Error("Cumulative weight overflow", zap.Uint64("cumulative", processingWeight), zap.Uint64("added", nw.Weight))
+				return false
+			}
+			processingWeight = newProcessingWeight
+		}
+	}
+
+	sd.log.Trace("Counted total weight that accepted blocks we're still processing", zap.Uint64("weight", processingWeight))
+
+	ratio := float64(processingWeight) / float64(totalValidatorWeight)
+
+	if ratio > stakeThresholdForStragglerSuspicion {
+		sd.log.Trace("We are straggling behind", zap.Float64("ratio", ratio))
+		return true
+	}
+
+	sd.log.Trace("Nodes ahead of us:", zap.Float64("ratio", ratio))
+
+	return false
+}
+
+func (sd *stragglerDetector) getNetworkSnapshot() (snapshot, bool) {
+	nodeWeightToLastAccepted, totalValidatorWeight := sd.getNetworkInfo()
+	if len(nodeWeightToLastAccepted) == 0 {
+		return snapshot{}, false
+	}
+
+	ourLastAcceptedBlock := sd.lastAccepted()
+
+	prevLastAcceptedCount := len(nodeWeightToLastAccepted)
+	for k, v := range nodeWeightToLastAccepted {
+		if ourLastAcceptedBlock.Compare(v) == 0 {
+			delete(nodeWeightToLastAccepted, k)
+		}
+	}
+	newLastAcceptedCount := len(nodeWeightToLastAccepted)
+
+	sd.log.Trace("Excluding nodes with our own height", zap.Int("prev", prevLastAcceptedCount), zap.Int("new", newLastAcceptedCount))
+
+	// Ensure we have collected last accepted blocks that are not our own last accepted block
+	// for at least 80% stake of the total weight we are connected to.
+
+	totalWeightWeKnowItsLastAcceptedBlock, err := nodeWeightToLastAccepted.totalWeight()
+	if err != nil {
+		sd.log.Error("Failed computing total weight", zap.Error(err))
+		return snapshot{}, false
+	}
+
+	ratio := float64(totalWeightWeKnowItsLastAcceptedBlock) / float64(totalValidatorWeight)
+
+	if ratio < 0.8 {
+		sd.log.Trace("Most stake we're connected to has the same height as we do",
+			zap.Float64("ratio", ratio))
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		nodeWeights2Blocks:   nodeWeightToLastAccepted,
+		totalValidatorWeight: totalValidatorWeight,
+	}
+
+	if sd.haveWeFailedCatchingUp(snap) {
+		return snap, true
+	}
+
+	return snapshot{}, false
+}
+
+func (sd *stragglerDetector) getNetworkInfo() (nodeWeights2Blocks, uint64) {
+	ratio := sd.connectedPercent()
+	if ratio < sd.minConfirmationThreshold {
+		// We don't know for sure whether we're behind or not.
+		// Even if we're behind, it's pointless to act before we have established
+		// connectivity with enough validators.
+		sd.log.Verbo("not enough connected stake to determine network info", zap.Float64("ratio", ratio))
+		return nil, 0
+	}
+
+	validators := nodeWeights(sd.connectedValidators().List())
+
+	nodeWeight2lastAccepted := make(nodeWeights2Blocks, len(validators))
```

Comment:
could we change the name of this variable as well eg `nodeWeightToLastAccepted`?